src/merkle_tree/incremental_merkle_tree.rs

```diff
@@ -0,0 +1,277 @@
+    pub fn current_index(&self) -> Option<usize> {
+        if self.is_empty() {
+            None
+        } else {
+            Some(self.current_path.leaf_index)
+        }
+    }
+
+    /// The next available index of leaf node
+    pub fn next_available(&self) -> Option<usize> {
+        let current_index = self.current_path.leaf_index;
+        if self.is_empty() {
+            Some(0)
+        } else if current_index < self.leaf_nodes.len() - 1 {
+            Some(current_index + 1)
+        } else {
+            None
+        }
+    }
+
+    /// Create an empty merkle tree such that all leaves are zero-filled.
+    pub fn blank(
+        leaf_hash_param: &LeafParam<P>,
+        two_to_one_hash_param: &TwoToOneParam<P>,
+        height: usize,
+    ) -> Result<Self, crate::Error> {
+        assert!(
+            height > 1,
+            "the height of incremental merkle tree should be at least 2"
+        );
+        // use empty leaf digest
+        let capacity: usize = 1 << (height - 1);
+        let leaves_digest = vec![P::LeafDigest::default(); capacity];
+        Ok(IncrementalMerkleTree {
+            /// blank tree doesn't have current_path
+            current_path: Path {
+                leaf_sibling_hash: P::LeafDigest::default(),
+                auth_path: Vec::new(),
+                leaf_index: 0,
+            },
+            leaf_nodes: leaves_digest,
+            two_to_one_hash_param: two_to_one_hash_param.clone(),
+            leaf_hash_param: leaf_hash_param.clone(),
+            root: P::InnerDigest::default(),
+            height,
+            empty: true,
+        })
+    }
+
+    /// Append leaf at `next_available`
+    /// ```tree_diagram
+    ///         [A]
+    ///        /   \
+    ///      [B]   ()
+    ///     / \   /  \
+    ///    D [E] ()  ()
+    ///   .. / \ ....
+    ///    [I]{new leaf}
+    /// ```
+    /// append({new leaf}) when the `next_availabe` is at 4, would cause a recompute [E], [A], [B]
+    pub fn append<T: Borrow<P::Leaf>>(&mut self, new_leaf: T) -> Result<(), crate::Error> {
+        assert!(self.next_available() != None, "index out of range");
+        let leaf_digest = P::LeafHash::evaluate(&self.leaf_hash_param, new_leaf)?;
+        let (path, root) = self.next_path(leaf_digest)?;
+        self.current_path = path;
+        self.root = root;
+        self.empty = false;
+        Ok(())
+    }
+
+    /// Generate updated path of `next_available` without changing the tree
+    /// returns (new_path, new_root)
+    pub fn next_path(
+        &self,
+        new_leaf_digest: P::LeafDigest,
+    ) -> Result<(Path<P>, P::InnerDigest), crate::Error> {
+        assert!(self.next_available() != None, "index out of range");
+
+        // calculate tree_height and empty hash
+        let tree_height = tree_height(self.leaf_nodes.len());
+        let hash_of_empty_node: P::InnerDigest = P::InnerDigest::default();
+        let hash_of_empty_leaf: P::LeafDigest = P::LeafDigest::default();
```

Comment:
I cannot think of any immediate harm here.

Reply:
got it thanks! looks fine

Reply:
If the intention is to be compatible with Zcash note commitment trees, then there's a sentinel value for empty leaves but there is no special case for empty internal nodes; they have the same value as would be expected from hashing their children. This doesn't require actually computing hashes for empty internal nodes, since the hash of a completely empty subtree only depends on its height (and can be computed in logarithmic time, then cached).

Reply:
@daira Thanks for your comments! Honestly I was just following the current arkwork's merkle tree implementation and didn't give too much thought. Now digging into it deeper, here is what I find:

For example, on `JubJub`, both `P::InnerDigest::default()` and `P::LeafDigest::default()` is `{x:0, y: <one of 0's corresponding y on the curve> }`. In the current implementation, the empty inner node a special value that is not equal to the `true` value where you actually compute from subtree.  Like you said, it is easy to compute these empty inner nodes once and then use the cached value.    

It is indeed not super "principled" but I am not too much concerned security wise due to the CRH (please let me know if i am wrong). If you think there is a concrete security issue here. We should fix both this PR and the arkworks merkle tree implementation.

Reply:
There is technically a small security issue in that if you use a sentinel value for internal nodes, you must argue that it was chosen such that it would be infeasible to find a preimage. (I.e. it is technically possible for there to be a back door where the sentinel is chosen to be the hash of some nonempty subtree.) But the main issue is just interoperability.